contrib/golang/upstreams/http/tcp/source/upstream_request.h

```diff
@@ -0,0 +1,214 @@
+    Success = 200,
+    InternalError = 500,
+  };
+  enum class EnvoyValue {
+    RouteName = 1,
+    ClusterName,
+  };
+
+  void initResponse();
+
+  // GenericUpstream
+  Envoy::Http::Status encodeHeaders(const Envoy::Http::RequestHeaderMap& headers, bool end_stream) override;
+  void encodeData(Buffer::Instance& data, bool end_stream) override;
+  void encodeMetadata(const Envoy::Http::MetadataMapVector&) override {}
+  void encodeTrailers(const Envoy::Http::RequestTrailerMap&) override;
+  void enableTcpTunneling() override {upstream_conn_data_->connection().enableHalfClose(true);};
+  void readDisable(bool disable) override;
+  void resetStream() override;
+  void setAccount(Buffer::BufferMemoryAccountSharedPtr) override {}
+
+  // Tcp::ConnectionPool::UpstreamCallbacks
+  void onUpstreamData(Buffer::Instance& data, bool end_stream) override;
+  void onEvent(Network::ConnectionEvent event) override;
+  void onAboveWriteBufferHighWatermark() override;
+  void onBelowWriteBufferLowWatermark() override;
+  const StreamInfo::BytesMeterSharedPtr& bytesMeter() override { return bytes_meter_; }
+
+  void trySendProxyData(bool send_data_to_upstream, bool end_stream);
+  void encodeDataGo(ProcessorState* state, Buffer::Instance& data, bool end_stream);
+  void sendDataToDownstream(Buffer::Instance& data, bool end_stream);
+
+  CAPIStatus copyHeaders(ProcessorState& state, GoString* go_strs, char* go_buf);
+  CAPIStatus setRespHeader(ProcessorState& state, absl::string_view key, absl::string_view value, headerAction act);
+  CAPIStatus removeRespHeader(ProcessorState& state, absl::string_view key);
+  CAPIStatus copyBuffer(ProcessorState& state, Buffer::Instance* buffer, char* data);
+  CAPIStatus drainBuffer(ProcessorState& state, Buffer::Instance* buffer, uint64_t length);
+  CAPIStatus setBufferHelper(ProcessorState& state, Buffer::Instance* buffer, absl::string_view& value, bufferAction action);
+  CAPIStatus getStringValue(int id, uint64_t* value_data, int* value_len);
+  CAPIStatus setSelfHalfCloseForUpstreamConn(int enabled);
+
+  DecodingProcessorState* decodingState() { return decoding_state_; }
+  EncodingProcessorState* encodingState() { return encoding_state_; }
+
+  EncodingProcessorState* encoding_state_;
+  DecodingProcessorState* decoding_state_;
+
+  const Router::RouteEntry* route_entry_;
+
+  // cache routeName for getStringValue,
+  // since upstream_request_->route().virtualHost().routeConfig().name() is not concurrent safe when c thread and go thread running at same time.
+  std::string route_name_;
+
+private:
+  Router::UpstreamToDownstream* upstream_request_;
+  Envoy::Tcp::ConnectionPool::ConnectionDataPtr upstream_conn_data_;
+  StreamInfo::BytesMeterSharedPtr bytes_meter_{std::make_shared<StreamInfo::BytesMeter>()};
+
+  Dso::TcpUpstreamDsoPtr dynamic_lib_;
+
+  bool already_send_resp_headers_{false};
+
+  // lock to avoid race between c thread and multi go threads (when c is running and calling back from go).
```

Comment:
Nope, in this case, Go code should only run in Envoy thread.

Reply:
okay, optimize the annotation.